core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -443,60 +449,25 @@ sealed abstract class Doc extends Product with Serializable {
     loop(Left(this) :: Nil, Doc.empty)
   }
 
-  /**
-   * This method is similar to flatten, but returns None if no
-   * flattening was needed (i.e. if no newlines or unions were present).
-   *
-   * As with flatten, the resulting Doc (if any) will never render any
-   * newlines, no matter what width is used.
-   */
-  def flattenOption: Option[Doc] = {
-    val res = flattenBoolean
-    if (res._2) Some(res._1) else None
-  }
-
-  /**
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenBoolean._1
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
+  private[paiges] def flatten: Doc = flattenBoolean._1
 
+  private[paiges] def flattenBoolean: (Doc, Boolean) = {
     type DB = (Doc, Boolean)
 
-    def finish(last: DB, front: List[DB]): DB =
-      front.foldLeft(last) {
-        case ((d1, c1), (d0, c2)) => (Concat(d0, d1), c1 || c2)
-      }
-
     @tailrec
-    def loop(h: DB, stack: List[DB], front: List[DB]): DB =
-      h._1 match {
-        case Empty | Text(_) =>
-          stack match {
-            case Nil => finish(h, front)
-            case x :: xs => loop(x, xs, h :: front)
-          }
-        case l@Line(_) =>
-          val next = l.asFlatDoc
-          val change = (next, true)
-          stack match {
-            case Nil => finish(change, front)
-            case x :: xs => loop(x, xs, change :: front)
-          }
-        case Nest(i, d) =>
-          loop((d, h._2), stack, front) // no Line, so Nest is irrelevant
-        case Align(d) =>
-          loop((d, h._2), stack, front) // no Line, so Align is irrelevant
-        case LazyDoc(d) => loop((d.evaluated, h._2), stack, front)
-        case Union(a, _) => loop((a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
-        case Concat(a, b) => loop((a, h._2), (b, h._2) :: stack, front)
+    def loop(h: DB, lst: List[DB], ks: List[DB => DB]): DB = h._1 match {
```

Comment:
I feel like this is more complex (using Fail) and I fear slower (since we have to use continuations).

What if we don't use Fail, which I think we only create here, and instead have `private def flatten: Option[Doc]` and `def flattenBoolean: Option[(Doc, Boolean)]`

back to the idea of 3 possible results here:
1. unflattenable `None`
2. flattenable, but no-op flatten `Some((d, false)`
3. flattenable, `Some((d, true)`

With this:
```scala
def group: Doc =
  flattenBoolean(this) match {
    case None => this
    case Some((d, flattened)) =>
      if (flattened) Union(d, this) else d
}
```
by the way, I guess due to our invariants we could also do:
```scala
def group: Doc =
  this match {
    case union@Union(_, _) => union // unions are already grouped
    case ...
```

Reply:
OK, I removed `Fail` and have more or less what you proposed.  The only wrinkle is in my testing of `fill` I needed some invariants on non-flattenable documents, so I write a `normalize` method that I apply to such documents.  I didn't see an obvious way to make it tail recursive, so I left it in the normal recursive style.  I figured if you're happy with the change, I can spend more effort to make it tail recursive.